File: Toy-Compiler/rust/src/type_checker.rs
//! Type checker for Zap. Walks the AST and enforces static types.

use std::collections::HashMap;
use crate::ast::*;

struct Scope {
    vars:   HashMap<String, ZapType>,
    parent: Option<Box<Scope>>,
}

impl Scope {
    fn new() -> Self { Self { vars: HashMap::new(), parent: None } }

    fn push(self) -> Self {
        Self { vars: HashMap::new(), parent: Some(Box::new(self)) }
    }

    fn pop(self) -> Self {
        *self.parent.expect("popped root scope")
    }

    fn define(&mut self, name: String, ty: ZapType) {
        self.vars.insert(name, ty);
    }

    fn lookup(&self, name: &str) -> Option<&ZapType> {
        self.vars.get(name).or_else(|| self.parent.as_ref()?.lookup(name))
    }
}

pub struct TypeChecker {
    fns:         HashMap<String, (Vec<ZapType>, ZapType)>,
    scope:       Scope,
    return_type: ZapType,
}

impl TypeChecker {
    pub fn new() -> Self {
        Self {
            fns:         HashMap::new(),
            scope:       Scope::new(),
            return_type: ZapType::Void,
        }
    }

    pub fn check(&mut self, program: &Program) -> Result<(), String> {
        for decl in &program.decls {
            if let Decl::Fn(fn_decl) = decl {
                let param_types: Vec<ZapType> = fn_decl.params.iter().map(|(_, t)| t.clone()).collect();
                self.fns.insert(fn_decl.name.clone(), (param_types, fn_decl.return_type.clone()));
            }
        }
        for decl in &program.decls {
            match decl {
                Decl::Fn(fn_decl) => self.check_fn(fn_decl)?,
                Decl::Stmt(stmt)  => self.check_stmt(stmt)?,
            }
        }
        Ok(())
    }

    fn enter(&mut self) {
        let old = std::mem::replace(&mut self.scope, Scope::new());
        self.scope = old.push();
    }

    fn leave(&mut self) {
        let old = std::mem::replace(&mut self.scope, Scope::new());
        self.scope = old.pop();
    }

    fn check_fn(&mut self, fn_decl: &FnDecl) -> Result<(), String> {
        let saved_ret = self.return_type.clone();
        self.return_type = fn_decl.return_type.clone();
        self.enter();
        for (name, ty) in &fn_decl.params {
            self.scope.define(name.clone(), ty.clone());
        }
        for stmt in &fn_decl.body.stmts { self.check_stmt(stmt)?; }
        self.leave();
        self.return_type = saved_ret;
        Ok(())
    }

    fn check_block(&mut self, block: &Block) -> Result<(), String> {
        self.enter();
        for stmt in &block.stmts { self.check_stmt(stmt)?; }
        self.leave();
        Ok(())
    }

    fn check_stmt(&mut self, stmt: &Stmt) -> Result<(), String> {
        match stmt {
            Stmt::Let(s) => {
                let got = self.check_expr(&s.value)?;
                if got != s.ty {
                    return Err(format!("Line {}: let {}: declared {}, got {}", s.line, s.name, s.ty, got));
                }
                self.scope.define(s.name.clone(), s.ty.clone());
            }
            Stmt::Return(s) => {
                let ty = match &s.value { None => ZapType::Void, Some(e) => self.check_expr(e)? };
                if ty != self.return_type {
                    return Err(format!("Line {}: return {} in {} function", s.line, ty, self.return_type));
                }
            }
            Stmt::If(s) => {
                let ct = self.check_expr(&s.cond)?;
                if ct != ZapType::Bool {
                    return Err(format!("Line {}: if condition must be bool, got {}", s.line, ct));
                }
                self.check_block(&s.then_block)?;
                if let Some(b) = &s.else_block { self.check_block(b)?; }
            }
            Stmt::While(s) => {
                let ct = self.check_expr(&s.cond)?;
                if ct != ZapType::Bool {
                    return Err(format!("Line {}: while condition must be bool, got {}", s.line, ct));
                }
                self.check_block(&s.body)?;
            }
            Stmt::Print(s) => { self.check_expr(&s.value)?; }
            Stmt::Expr(s)  => { self.check_expr(&s.expr)?; }
        }
        Ok(())
    }

    fn check_expr(&mut self, expr: &Expr) -> Result<ZapType, String> {
        match expr {
            Expr::IntLit(_, _)  => Ok(ZapType::Int),
            Expr::BoolLit(_, _) => Ok(ZapType::Bool),
            Expr::Ident(name, line) => {
                self.scope.lookup(name).cloned()
                    .ok_or_else(|| format!("Line {line}: undefined variable '{name}'"))
            }
            Expr::Assign { name, value, line } => {
                let existing = self.scope.lookup(name).cloned()
                    .ok_or_else(|| format!("Line {line}: undefined variable '{name}'"))?;
                let got = self.check_expr(value)?;
                if got != existing {
                    return Err(format!("Line {line}: cannot assign {got} to {existing} variable '{name}'"));
                }
                Ok(existing)
            }
            Expr::BinOp { op, left, right, line } => {
                let lt = self.check_expr(left)?;
                let rt = self.check_expr(right)?;
                match op.as_str() {
                    "+" | "-" | "*" | "/" => {
                        if lt != ZapType::Int || rt != ZapType::Int {
                            return Err(format!("Line {line}: {op} requires int operands, got {lt} and {rt}"));
                        }
                        Ok(ZapType::Int)
                    }
                    "<" | ">" | "<=" | ">=" => {
                        if lt != ZapType::Int || rt != ZapType::Int {
                            return Err(format!("Line {line}: {op} requires int operands"));
                        }
                        Ok(ZapType::Bool)
                    }
                    "==" | "!=" => {
                        if lt != rt {
                            return Err(format!("Line {line}: {op} requires matching types, got {lt} and {rt}"));
                        }
                        Ok(ZapType::Bool)
                    }
                    "&&" | "||" => {
                        if lt != ZapType::Bool || rt != ZapType::Bool {
                            return Err(format!("Line {line}: {op} requires bool operands"));
                        }
                        Ok(ZapType::Bool)
                    }
                    _ => Err(format!("Unknown binary operator '{op}'")),
                }
            }
            Expr::UnaryOp { op, operand, line } => {
                let ty = self.check_expr(operand)?;
                match op.as_str() {
                    "!" => {
                        if ty != ZapType::Bool { return Err(format!("Line {line}: '!' requires bool, got {ty}")); }
                        Ok(ZapType::Bool)
                    }
                    "-" => {
                        if ty != ZapType::Int { return Err(format!("Line {line}: unary '-' requires int, got {ty}")); }
                        Ok(ZapType::Int)
                    }
                    _ => Err(format!("Unknown unary operator '{op}'")),
                }
            }
            Expr::Call { name, args, line } => {
                let (param_types, ret) = self.fns.get(name).cloned()
                    .ok_or_else(|| format!("Line {line}: undefined function '{name}'"))?;
                if args.len() != param_types.len() {
                    return Err(format!("Line {line}: '{name}' expects {} args, got {}", param_types.len(), args.len()));
                }
                for (i, (arg, expected)) in args.iter().zip(param_types.iter()).enumerate() {
                    let got = self.check_expr(arg)?;
                    if got != *expected {
                        return Err(format!("Line {line}: arg {i} of '{name}': expected {expected}, got {got}"));
                    }
                }
                Ok(ret)
            }
        }
    }
}

pub fn type_check(program: &Program) -> Result<(), String> {
    TypeChecker::new().check(program)
}

File: modules/compiler/python/zap/type_checker.py
"""Type checker for Zap. Walks the AST and enforces static types."""

from __future__ import annotations
from .ast_nodes import *


class TypeErrorZap(Exception):
    pass


class Scope:
    def __init__(self, parent: Scope | None = None) -> None:
        self._vars: dict[str, str] = {}
        self.parent = parent

    def define(self, name: str, ty: str) -> None:
        self._vars[name] = ty

    def lookup(self, name: str) -> str | None:
        if name in self._vars:
            return self._vars[name]
        return self.parent.lookup(name) if self.parent else None


class TypeChecker:
    def __init__(self) -> None:
        # fn_name -> ([param_types], return_type)
        self._fns: dict[str, tuple[list[str], str]] = {}
        self._scope = Scope()
        self._return_type = "void"

    def check(self, program: Program) -> None:
        # Pass 1: collect all function signatures so calls can be forward-referenced.
        for decl in program.decls:
            if isinstance(decl, FnDecl):
                self._fns[decl.name] = (
                    [ty for _, ty in decl.params],
                    decl.return_type,
                )
        # Pass 2: check bodies.
        for decl in program.decls:
            if isinstance(decl, FnDecl):
                self._check_fn(decl)
            else:
                self._check_stmt(decl)

    # ── functions ─────────────────────────────────────────────────────────────

    def _check_fn(self, fn: FnDecl) -> None:
        saved_ret = self._return_type
        self._return_type = fn.return_type
        self._scope = Scope(self._scope)
        for name, ty in fn.params:
            self._scope.define(name, ty)
        for stmt in fn.body.stmts:
            self._check_stmt(stmt)
        self._scope = self._scope.parent
        self._return_type = saved_ret

    def _check_block(self, block: Block) -> None:
        self._scope = Scope(self._scope)
        for stmt in block.stmts:
            self._check_stmt(stmt)
        self._scope = self._scope.parent

    # ── statements ────────────────────────────────────────────────────────────

    def _check_stmt(self, stmt: Stmt) -> None:
        if isinstance(stmt, LetStmt):
            got = self._check_expr(stmt.value)
            if got != stmt.ty:
                raise TypeErrorZap(
                    f"Line {stmt.line}: let {stmt.name}: declared {stmt.ty}, got {got}"
                )
            self._scope.define(stmt.name, stmt.ty)

        elif isinstance(stmt, ReturnStmt):
            ty = "void" if stmt.value is None else self._check_expr(stmt.value)
            if ty != self._return_type:
                raise TypeErrorZap(
                    f"Line {stmt.line}: return {ty} in {self._return_type!r} function"
                )

        elif isinstance(stmt, IfStmt):
            ct = self._check_expr(stmt.cond)
            if ct != "bool":
                raise TypeErrorZap(f"Line {stmt.line}: if condition must be bool, got {ct}")
            self._check_block(stmt.then_block)
            if stmt.else_block:
                self._check_block(stmt.else_block)

        elif isinstance(stmt, WhileStmt):
            ct = self._check_expr(stmt.cond)
            if ct != "bool":
                raise TypeErrorZap(f"Line {stmt.line}: while condition must be bool, got {ct}")
            self._check_block(stmt.body)

        elif isinstance(stmt, PrintStmt):
            self._check_expr(stmt.value)

        elif isinstance(stmt, ExprStmt):
            self._check_expr(stmt.expr)

    # ── expressions ───────────────────────────────────────────────────────────

    def _check_expr(self, expr: Expr) -> str:
        if isinstance(expr, IntLit):
            return "int"
        if isinstance(expr, BoolLit):
            return "bool"
        if isinstance(expr, IdentExpr):
            ty = self._scope.lookup(expr.name)
            if ty is None:
                raise TypeErrorZap(f"Line {expr.line}: undefined variable '{expr.name}'")
            return ty
        if isinstance(expr, AssignExpr):
            existing = self._scope.lookup(expr.name)
            if existing is None:
                raise TypeErrorZap(f"Line {expr.line}: undefined variable '{expr.name}'")
            got = self._check_expr(expr.value)
            if existing != got:
                raise TypeErrorZap(
                    f"Line {expr.line}: cannot assign {got} to {existing} variable '{expr.name}'"
                )
            return existing
        if isinstance(expr, BinOp):
            return self._check_binop(expr)
        if isinstance(expr, UnaryOp):
            return self._check_unary(expr)
        if isinstance(expr, CallExpr):
            return self._check_call(expr)
        raise TypeErrorZap(f"Unknown expression kind: {type(expr).__name__}")

    def _check_binop(self, e: BinOp) -> str:
        lt = self._check_expr(e.left)
        rt = self._check_expr(e.right)
        if e.op in ("+", "-", "*", "/"):
            if lt != "int" or rt != "int":
                raise TypeErrorZap(f"Line {e.line}: {e.op} requires int operands, got {lt} and {rt}")
            return "int"
        if e.op in ("<", ">", "<=", ">="):
            if lt != "int" or rt != "int":
                raise TypeErrorZap(f"Line {e.line}: {e.op} requires int operands")
            return "bool"
        if e.op in ("==", "!="):
            if lt != rt:
                raise TypeErrorZap(f"Line {e.line}: {e.op} requires matching types, got {lt} and {rt}")
            return "bool"
        if e.op in ("&&", "||"):
            if lt != "bool" or rt != "bool":
                raise TypeErrorZap(f"Line {e.line}: {e.op} requires bool operands")
            return "bool"
        raise TypeErrorZap(f"Unknown binary operator '{e.op}'")

    def _check_unary(self, e: UnaryOp) -> str:
        ty = self._check_expr(e.operand)
        if e.op == "!":
            if ty != "bool":
                raise TypeErrorZap(f"Line {e.line}: '!' requires bool operand, got {ty}")
            return "bool"
        if e.op == "-":
            if ty != "int":
                raise TypeErrorZap(f"Line {e.line}: unary '-' requires int operand, got {ty}")
            return "int"
        raise TypeErrorZap(f"Unknown unary operator '{e.op}'")

    def _check_call(self, e: CallExpr) -> str:
        if e.name not in self._fns:
            raise TypeErrorZap(f"Line {e.line}: undefined function '{e.name}'")
        param_types, ret = self._fns[e.name]
        if len(e.args) != len(param_types):
            raise TypeErrorZap(
                f"Line {e.line}: '{e.name}' expects {len(param_types)} args, got {len(e.args)}"
            )
        for i, (arg, expected) in enumerate(zip(e.args, param_types)):
            got = self._check_expr(arg)
            if got != expected:
                raise TypeErrorZap(
                    f"Line {e.line}: arg {i} of '{e.name}': expected {expected}, got {got}"
                )
        return ret


def type_check(program: Program) -> None:
    TypeChecker().check(program)
